content/blog/mixtral-8x7b-healthcare-chatbot/index.md

```diff
@@ -0,0 +1,167 @@
+- The human impact of providing wrong answers. 😬
+- *Staying up to date*. It takes a long time to train a LLM. By the time they are released, their knowledge is already a bit stale. Medical research breakthroughs are happening all the time. I am particularly interested in long COVID, which has both impacted humanity on a massive scale, and is the subject of ongoing research.
+
+Therefore, I decided to use a RAG pipeline to combine PubMed data with a LLM.
+
+## The PubMed RAG Pipeline
+
+Retrieval augmented generation, or RAG, is a way of giving a LLM context so it can better answer questions. 
+
+You pass the LLM some documents, along with a query, and prompt the LLM to use the documents when answering the question.
+
+![](RAG.png)
+
+[PubMed](https://pubmed.ncbi.nlm.nih.gov/) has up to date, trustworthy medical information so it seemed like a solid document source. Plus, there's a [PyMed wrapper for the PubMed API](https://github.com/gijswobben/pymed) that made querying easy peasy. We'll wrap this in a [Haystack custom component](https://docs.haystack.deepset.ai/v2.0/docs/custom-components) to format the results as `Document`s so that Haystack can use them, and add some light error handling.
+
+```python
+from pymed import PubMed
+from typing import List
+from haystack import component
+from haystack import Document
+
+pubmed = PubMed(tool="Haystack2.0Prototype", email="tilde.thurium@deepset.ai")
+
+def documentize(article):
+  return Document(content=article.abstract, meta={'title': article.title, 'keywords': article.keywords})
+
+@component
+class PubMedFetcher():
+
+  @component.output_types(articles=List[Document])
+  def run(self, queries: list[str]):
+    cleaned_queries = queries[0].strip().split('\n')
+
+    articles = []
+    try:
+      for query in cleaned_queries:
+        response = pubmed.query(query, max_results = 1)
+        documents = [documentize(article) for article in response]
+        articles.extend(documents)
+    except Exception as e:
+        print(e)
+        print(f"Couldn't fetch articles for queries: {queries}" )
+    results = {'articles': articles}
+    return results
+```
+
+For the model, I went with Mixtral's 8x7b. Mixtral is a unique kind of model that uses 8 *"experts"* and an internal *"routing"* mechanism that routes a token to a specific expert. This also means that during inferencing, not all parameters are used, which allows the model to response remarkably fast.  [This HuggingFace blog post explains MoE](https://huggingface.co/blog/moe) in more detail. 
+
+
+## Generating Keywords for PubMed with Mixtral/LLMs
+
+First, I tried an approach where I passed a plain query to PubMed. e.g. *"What are the most  current treatments for long COVID?"* Unfortunately, that didn't work too well. The articles returned weren't very relevant. Which makes sense, because PubMed isnt optimized for natural language search. It is optimized for keywords, though. And you know what's great at generating keywords? LLMs!
```

Comment:
Loved this section, just a small typo here
```suggestion
First, I tried an approach where I passed a plain query to PubMed. e.g. *"What are the most  current treatments for long COVID?"* Unfortunately, that didn't work too well. The articles returned weren't very relevant. Which makes sense, because PubMed isn't optimized for natural language search. It is optimized for keywords, though. And you know what's great at generating keywords? LLMs!
```